core/client/src/main/java/alluxio/client/block/stream/PacketOutStream.java

```diff
@@ -0,0 +1,284 @@
+ * streams data packet by packet.
+ */
+@NotThreadSafe
+public final class PacketOutStream extends OutputStream implements BoundedStream, Cancelable {
+  /** Length of the stream. If unknown, set to Long.MAX_VALUE. */
+  private final long mLength;
+  private ByteBuf mCurrentPacket = null;
+
+  private final List<PacketWriter> mPacketWriters;
+  private boolean mClosed;
+
+  /**
+   * Creates a {@link PacketOutStream} that writes to a local file.
+   *
+   * @param client the block worker client
+   * @param id the ID
+   * @param length the block or file length
+   * @return the {@link PacketOutStream} created
+   * @throws IOException if it fails to create the object
+   */
+  public static PacketOutStream createLocalPacketOutStream(BlockWorkerClient client,
+      long id, long length) throws IOException {
+    PacketWriter packetWriter = LocalFilePacketWriter.create(client, id);
+    return new PacketOutStream(packetWriter, length);
+  }
+
+  /**
+   * Creates a {@link PacketOutStream} that writes to a netty data server.
+   *
+   * @param context the file system context
+   * @param address the netty data server address
+   * @param sessionId the session ID
+   * @param id the ID (block ID or UFS file ID)
+   * @param length the block or file length
+   * @param type the request type (either block write or UFS file write)
+   * @return the {@link PacketOutStream} created
+   * @throws IOException if it fails to create the object
+   */
+  public static PacketOutStream createNettyPacketOutStream(FileSystemContext context,
+      InetSocketAddress address, long sessionId, long id, long length, Protocol.RequestType type)
+      throws IOException {
+    NettyPacketWriter packetWriter =
+        new NettyPacketWriter(context, address, id, length, sessionId, type);
+    return new PacketOutStream(packetWriter, length);
+  }
+
+  /**
+   * Creates a {@link PacketOutStream} that writes to a list of locations.
+   *
+   * @param context the file system context
+   * @param clients the netty data server addresses
+   * @param sessionIds the session IDs for all the data servers
+   * @param id the ID (block ID or UFS file ID)
+   * @param length the block or file length
+   * @param type the request type (either block write or UFS file write)
+   * @return the {@link PacketOutStream} created
+   * @throws IOException if it fails to create the object
+   */
+  public static PacketOutStream createReplicatedPacketOutStream(FileSystemContext context,
+      List<BlockWorkerClient> clients, List<Long> sessionIds, long id, long length,
+      Protocol.RequestType type) throws IOException {
+    Preconditions.checkArgument(clients.size() == sessionIds.size());
+    String localHost = NetworkAddressUtils.getLocalHostName();
+
+    Iterator<Long> iterator = sessionIds.iterator();
+    List<PacketWriter> packetWriters = new ArrayList<>();
+    for (BlockWorkerClient client: clients) {
```

Comment:
This seems a bit clunky, since you are essentially trying to iterate over both clients and session ids. Maybe add a TODO to clean this up?